File: src/domain/enemy.ts
/**
 * Enemies.
 *
 * Enemies carry a flat stat block rather than a growth curve - encounter
 * design is easier when a slime's numbers are written down instead of derived.
 * Behaviour is a weighted list of moves, which is enough for most encounters;
 * a boss that needs phases supplies `behaviorTag` and the project's own
 * scripting layer takes over.
 */

import type { Element } from './elements.js';
import type { StatBlock } from './stats.js';

export interface EnemyMove {
  /** Art id, or the sentinel "attack" for a basic strike. */
  artId: string;
  weight: number;
  /** Only considered when the enemy is at or below this HP fraction. */
  hpBelow?: number;
  /** Only considered from this round onward. */
  fromRound?: number;
  /** Not usable again for this many rounds. */
  cooldown?: number;
}

export interface EnemyDrop {
  itemId: string;
  chance: number;
}

export interface EnemyDef {
  id: string;
  name: string;
  innate: Element;
  stats: StatBlock;
  arts?: string[];
  moves?: EnemyMove[];
  xp: number;
  coin: number;
  drops?: EnemyDrop[];
  spriteId?: string;
  /** Hook for project-specific boss scripting. */
  behaviorTag?: string;
  /** Resistance to being fled from, 0..1. */
  fleeResistance?: number;
}

export interface EncounterMember {
  enemyId: string;
  /** Formation slot; adjacency drives spread damage. */
  slot: number;
}

export interface EncounterDef {
  id: string;
  members: EncounterMember[];
  /** Cannot be fled. */
  noFlee?: boolean;
  musicId?: string;
  backgroundId?: string;
}

File: unity/Aetherlight.Core/Domain/Enemy.cs
using System.Collections.Generic;

namespace Aetherlight.Domain
{
    /// <summary>
    /// Enemies carry a flat stat block rather than a growth curve - encounter
    /// design is easier when a slime's numbers are written down instead of
    /// derived.
    /// </summary>
    public sealed class EnemyMove
    {
        /// <summary>Art id, or the sentinel "attack" for a basic strike.</summary>
        public string ArtId = "";
        public double Weight;

        /// <summary>Only considered at or below this HP fraction.</summary>
        public double? HpBelow;

        /// <summary>Only considered from this round onward.</summary>
        public int? FromRound;

        /// <summary>Not usable again for this many rounds.</summary>
        public int Cooldown;
    }

    public sealed class EnemyDrop
    {
        public string ItemId = "";
        public double Chance;
    }

    public sealed class EnemyDef
    {
        public string Id = "";
        public string Name = "";
        public Element Innate;
        public StatBlock Stats = new StatBlock();
        public List<string> Arts = new List<string>();
        public List<EnemyMove> Moves = new List<EnemyMove>();
        public long Xp;
        public long Coin;
        public List<EnemyDrop> Drops = new List<EnemyDrop>();
        public string? SpriteId;

        /// <summary>Hook for project-specific boss scripting.</summary>
        public string? BehaviorTag;

        /// <summary>Resistance to being fled from, 0..1.</summary>
        public double FleeResistance;
    }

    public sealed class EncounterMember
    {
        public string EnemyId = "";
        /// <summary>Formation slot; adjacency drives spread damage.</summary>
        public int Slot;
    }

    public sealed class EncounterDef
    {
        public string Id = "";
        public List<EncounterMember> Members = new List<EncounterMember>();
        public bool NoFlee;
        public string? MusicId;
        public string? BackgroundId;
    }
}
